### DISCUSSION TITLE
EventRateLimit Admission control plugin issue

### DISCUSSION THREAD

**Discussion by bheemareddy** [k3s-io/k3s#5583] [Q&A]
Hi, I'm trying to add an admission plugin to k3s.
https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#eventratelimit

I'm facing an issue while applying limits to the event rate limit plugin.

```
[root@appliance-1 ~]# cat /etc/kubernetes/admission.yaml 
apiVersion: apiserver.k8s.io/v1alpha1
kind: AdmissionConfiguration
plugins:
- name: EventRateLimit
  path: /etc/kubernetes/event.yaml
[root@appliance-1 ~]# 
[root@appliance-1 ~]# cat /etc/kubernetes/event.yaml 
apiVersion: eventratelimit.admission.k8s.io/v1alpha1
kind: Configuration
limits:
- type: Namespace
  qps: 1
  burst: 10
  cacheSize: 2
- type: User
  qps: 1
  burst: 5
- type: Server
  qps: 1
  burst: 5

[root@appliance-1 ~]# kubectl apply -f /etc/kubernetes/admission.yaml 
error: unable to recognize "/etc/kubernetes/admission.yaml": no matches for kind "AdmissionConfiguration" in version "apiserver.k8s.io/v1alpha1"
[root@appliance-1 ~]# 
[root@appliance-1 ~]# 
[root@appliance-1 ~]# kubectl apply -f /etc/kubernetes/event.ayml
error: the path "/etc/kubernetes/event.ayml" does not exist
[root@appliance-1 ~]# kubectl apply -f /etc/kubernetes/event.yaml
error: unable to recognize "/etc/kubernetes/event.yaml": no matches for kind "Configuration" in version "eventratelimit.admission.k8s.io/v1alpha1"

```

Thanks in Advance

**Comment by brandond** [comment]
As noted in the docs you linked to, `EventRateLimit` is still in alpha which means that it's not enabled by default. Did you enable the featuregates on the apiserver to turn it on? Did you enable the admission controller? I suspect that without enabling it the resource types aren't even registered.

The docs also mention that you need to point the apiserver at that file with `--admission-control-config-file`. It is not something that you can apply to the cluster after it's up. Maybe take a second look at those docs?

**Reply by bheemareddy** [reply]
Thanks for the Info. I've followed the documents closely and I've added the plugin "EventRateLimit".  

I've added arguments for Kube API Server.

'--kube-apiserver-arg=enable-admission-plugins=EventRateLimit'
'--kube-apiserver-arg=admission-control-config-file=/etc/kubernetes/admission.yaml'

Later i've restarted the k3s service

systemctl daemon-reload
systemctl restart k3s.service

I see k3s restarted perfectly fine and journal logs saying the plugin is being added.


Now i do not see EventRateLimit in  "kubectl api-resources"
And even functionality also not working, Even if I send exceeding rest calls it is not blocking?

How to make sure it is applied to the API-Server or not?

**Reply by brandond** [reply]
I'm not sure if it is expected that the type would be listed in api-resources, but after re-reading the documentation I suspect not, as it's only used for configuration, not for resources.

This is an alpha feature and I haven't played with it myself, what behavior are you expecting from it and why?

**Reply by bheemareddy** [reply]
Hi Brandon,

Thanks for the info,

Our expectation is to control the number of requests served by the Kube API server.
As per the security requirement,
Threat
A misbehaving workload could overwhelm and lead to Denial Of Service for the API Server, right now, there is no known attack path. But if the attacker cracks the API Server IP Address and can easily bring down the complete cluster

**Reply by lazzio7** [reply]
Hi bheemareddy,

have you fix this issue ?
im in same situation regarding PSA security report...

```
rancher-03:/var/log/containers # cat /var/log/pods/kube-system_kube-apiserver-rancher-03_4874a08227e8932676b83ca998a390f3/kube-apiserver/1058.log
2023-09-27T10:34:10.366968508+02:00 stderr F I0927 08:34:10.366647       1 server.go:563] external host was not specified, using 192.168.47.116
2023-09-27T10:34:10.367201739+02:00 stderr F I0927 08:34:10.367094       1 server.go:161] Version: v1.25.9+rke2r1
2023-09-27T10:34:10.367215213+02:00 stderr F I0927 08:34:10.367126       1 server.go:163] "Golang settings" GOGC="" GOMAXPROCS="" GOTRACEBACK=""
2023-09-27T10:34:10.673893439+02:00 stderr F I0927 08:34:10.673741       1 shared_informer.go:252] Waiting for caches to sync for node_authorizer
2023-09-27T10:34:10.674649831+02:00 stderr F E0927 08:34:10.674403       1 run.go:74] "command failed" err="failed to initialize admission: failed to read plugin config: no kind \"Configuration\" is registered for version \"eventratelimit.admission.k8s.io/v1alpha1\" in scheme \"k8s.io/apiserver/pkg/server/options/admission.go:44\""
```

Thanks

**Reply by brandond** [reply]
@lazzio7 please read the comment you're replying to:
> `EventRateLimit` is still in alpha which means that it's not enabled by default. Did you enable the admission controller? I suspect that without enabling it the resource types aren't even registered.

Please post the complete RKE2 configuration that you're using on this node.

**Reply by lazzio7** [reply]
Hi brandond,

well, i have followed     [this](https://github.com/rancher/rke2/issues/930#issuecomment-916636460) approach and it works now,
question is, wheter is ok to implement into the production if it an alpha.

Thank You

**Reply by brandond** [reply]
I'm glad you got it working! Whether or not you feel comfortable using an Alpha feature in production is probably a question you'll need to answer for yourself.

**Reply by lazzio7** [reply]
:-) 
Thanks
